src/client/datascience/jupyter/jupyterServer.ts

```diff
@@ -61,19 +65,63 @@ export class JupyterServerBase implements INotebookServer {
                 this.shutdown().ignoreErrors();
             });
         }
+
+        // Create our session manager
+        this.sessionManager = await this.sessionManagerFactory.create(launchInfo.connectionInfo);
+
+        // Try creating a session just to ensure we're connected. Callers of this function check to make sure jupyter
+        // is running and connectable.
+        let session: IJupyterSession | undefined;
+        session = await this.sessionManager.startNew(launchInfo.kernelSpec, cancelToken);
+        const idleTimeout = this.configService.getSettings().datascience.jupyterLaunchTimeout;
+        // The wait for idle should throw if we can't connect.
+        await session.waitForIdle(idleTimeout);
+
+        // If that works, save this session for the next notebook to use
+        this.savedSession = session;
     }
 
     public createNotebook(resource: Uri, cancelToken?: CancellationToken): Promise<INotebook> {
-        return this.createNotebookInstance(resource, this.sessionManager, this.disposableRegistry, this.configService, this.loggers, cancelToken);
+        if (!this.sessionManager) {
+            throw new Error(localize.DataScience.sessionDisposed());
+        }
+        // If we have a saved session send this into the notebook so we don't create a new one
+        const savedSession = this.savedSession;
+        this.savedSession = undefined;
+
+        // Create a notebook and return it.
+        return this.createNotebookInstance(resource, this.sessionManager, savedSession, this.disposableRegistry, this.configService, this.loggers, cancelToken);
     }
 
     public async shutdown(): Promise<void> {
+        // Order should be
+        // 1) connectionInfoDisconnectHandler - listens to process close
+        // 2) sessions (owned by the notebooks)
+        // 3) session manager (owned by this object)
+        // 4) connInfo (owned by this object) - kills the jupyter process
+
         if (this.connectionInfoDisconnectHandler) {
             this.connectionInfoDisconnectHandler.dispose();
             this.connectionInfoDisconnectHandler = undefined;
         }
-        traceInfo(`Shutting down ${this.id}`);
+
+        // Destroy the kernel spec

```

Comment:
Forgot to eliminate the savedSession here. It should go away if no notebooks are created.